### PR TITLE
[Enhancement] Support Skipping compaction lower replica where select queryable replica for better scan performance

### LINKED ISSUE #8145: [Enhancement] Support Skipping compaction lower replica where select queryable replica
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Some olap table may receive realtime data, and some replicas may have larger version_count than others because of slower compaction, which will cause poor scan performance, so we give a optimize solution that skip compaction lower replica where select queryable replica, which will improve olap scan performance when we skip replica with obvious larger  versioncount.

### Solution

Support Skipping compaction lower replica where select queryable replica

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by caiconghui**
# Proposed changes

Issue Number: close #8145 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (No)
3. Has document been added or modified: (Yes)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (Yes)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java [hunk 1]
```diff
@@ -227,10 +228,25 @@ public void getQueryableReplicas(List<Replica> allQuerableReplica, long visibleV
...
-                    allQuerableReplica.add(replica);
+                    allQueryableReplica.add(replica);
                 }
             }
         }
+
+        if (Config.skip_compaction_slower_replica && allQueryableReplica.size() > 1) {
+            long minVersionCount = Long.MAX_VALUE;
+            for (Replica replica : allQueryableReplica) {
+                if (replica.getVersionCount() != -1 && replica.getVersionCount() < minVersionCount) {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java:239]: check replica.getVersionCount() > 0 is better. Since the tablet has at least one rowset

File: be/src/olap/tablet_manager.cpp [hunk 2]
```diff
@@ -888,19 +880,18 @@ OLAPStatus TabletManager::build_all_report_tablets_info(
 
     DorisMetrics::instance()->report_all_tablets_requests_total->increment(1);
     HistogramStat tablet_version_num_hist;
+    std::shared_ptr<std::vector<TTabletStat>> local_cache = std::make_shared<std::vector<TTabletStat>>();
```
Comment (levy5307) [inline be/src/olap/tablet_manager.cpp:883]: auto local_cache = std::make_shared<std::vector<TTabletStat>>();

File: be/src/olap/tablet_manager.cpp [hunk 3]
```diff
@@ -616,19 +615,12 @@ bool TabletManager::get_rowset_id_from_path(const string& path, RowsetId* rowset
 }
 
 void TabletManager::get_tablet_stat(TTabletStatResult* result) {
-    int64_t curr_ms = UnixMillis();
-    // Update cache if it is too old
+    std::shared_ptr<std::vector<TTabletStat>> local_cache;
```
Comment (levy5307) [inline be/src/olap/tablet_manager.cpp:618]: Why did you remove the update logic?

Comment (caiconghui) [reply be/src/olap/tablet_manager.cpp:618]: it is Inefficient，we can produce cache in build_tablet_info function

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Replica.java [hunk 4]
```diff
@@ -239,12 +239,19 @@ public void setNeedFurtherRepair(boolean needFurtherRepair) {
...
     }
 
-    // only update data size and row num
+    // for compatibility
     public synchronized void updateStat(long dataSize, long rowNum) {
         this.dataSize = dataSize;
         this.rowCount = rowNum;
     }
 
+    // only update data_size, row_num, version_count
```
Comment (levy5307) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Replica.java:248]: I think this comment is useless. Because we can get these information from function parameters

File: be/src/olap/tablet_manager.cpp [hunk 5]
```diff
@@ -913,12 +904,22 @@ OLAPStatus TabletManager::build_all_report_tablets_info(
...
-
             if (!t_tablet.tablet_infos.empty()) {
                 tablets_info->emplace(tablet_id, t_tablet);
+                TTabletStat t_tablet_stat;
+                const auto& tablet_info = t_tablet.tablet_infos[0];
+                t_tablet_stat.tablet_id = tablet_info.tablet_id;
+                t_tablet_stat.data_size = tablet_info.data_size;
+                t_tablet_stat.row_num = tablet_info.row_count;
+                t_tablet_stat.version_count = tablet_info.version_count;
+                local_cache->emplace_back(t_tablet_stat);
```
Comment (levy5307) [inline be/src/olap/tablet_manager.cpp:915]: use move constructor
```
local_cache->emplace_back(std::move(t_tablet_stat));
```

File: be/src/olap/tablet_manager.h [hunk 6]
```diff
@@ -228,12 +226,8 @@ class TabletManager {
...
     std::vector<TabletSharedPtr> _shutdown_tablets;
 
-    std::mutex _tablet_stat_mutex;
-    // cache to save tablets' statistics, such as data-size and row-count
-    // TODO(cmy): for now, this is a naive implementation
-    std::map<int64_t, TTabletStat> _tablet_stat_cache;
-    // last update time of tablet stat cache
-    int64_t _last_update_stat_ms;
+    std::mutex _tablet_stat_cache_mutex;
+    std::shared_ptr<std::vector<TTabletStat>> _tablet_stat_list_cache = std::make_shared<std::vector<TTabletStat>>();
```
Comment (levy5307) [inline be/src/olap/tablet_manager.h:230]: What about std::unique_ptr? It is more efficient than std::shared_ptr

Comment (caiconghui) [reply be/src/olap/tablet_manager.h:230]: no，the function will be called more than once

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java [hunk 7]
```diff
@@ -227,10 +228,25 @@ public void getQueryableReplicas(List<Replica> allQuerableReplica, long visibleV
...
+                    minVersionCount = replica.getVersionCount();
+                }
+            }
+            final long finalMinVersionCount = minVersionCount;
+            return allQueryableReplica.stream().filter(replica -> replica.getVersionCount() == -1 ||
+                            replica.getVersionCount() < Config.min_version_count_indicate_replica_compaction_too_slow ||
+                            replica.getVersionCount() < finalMinVersionCount * 3)
```
Comment (levy5307) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java:246]: What does the const var 3 mean?
Could you add a comment for it?

**Review by levy5307 — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java [hunk 8]
```diff
@@ -227,10 +231,25 @@ public void getQueryableReplicas(List<Replica> allQuerableReplica, long visibleV
...
         }
+
+        if (Config.skip_compaction_slower_replica && allQueryableReplica.size() > 1) {
+            long minVersionCount = Long.MAX_VALUE;
+            for (Replica replica : allQueryableReplica) {
+                if (replica.getVersionCount() != -1 && replica.getVersionCount() < minVersionCount) {
+                    minVersionCount = replica.getVersionCount();
+                }
+            }
+            final long finalMinVersionCount = minVersionCount;
```
Comment (weizuo93) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java:246]: Is this code redundant?

Comment (caiconghui) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Tablet.java:246]: lamda expression need final variables

**Review by weizuo93 — APPROVED** [review]
LGTM
